File: go/0134_gas_station.go
func canCompleteCircuit(gas []int, cost []int) int {
	glen := len(gas)
	var start, amount int
LOOP:
	// loop to set a new start point
	for i := 0; i < glen; {
		start, amount = i, 0
		// loop to check if we can complete a cycle from start to end
		for j := 0; j < glen; j++ {
			cur := (start + j) % glen
			if amount = amount + gas[cur] - cost[cur]; amount < 0 {
				if i+j >= glen {
					return -1 // all start points failed
				}
				i += j + 1 // skip the start point
				continue LOOP
			}
		}
		return start
	}
	return -1
}

File: python/0134_gas_station.py
class Solution:
    def canCompleteCircuit(self, gas, cost):
        """
        :type gas: List[int]
        :type cost: List[int]
        :rtype: int
        """

        i = 0
        while i < len(gas):
            amount = 0
            j = 0
            ok = True
            while j < len(gas):
                cur = (i + j) % len(gas)
                amount += gas[cur] - cost[cur]
                if amount < 0:
                    i = i + 1 if j == 0 else i + j
                    ok = False
                    break
                j += 1
            if not ok:
                continue
            return i
        return -1
